Scc 2898

charmingduchess opened the pull request.

mwbenowitz commented on lib/sierra_record.rb:
```diff
@@ -1,44 +1,41 @@
+      begin
+        sierra_record.encode_and_send_to_kinesis
+        $logger.info("Sent record to kinesis stream record ##{record['id']}")
+      rescue AvroError => e
+        $logger.warn("Record (id# #{record['id']} failed avro validation", { status: e.message })
+        @process_statuses[:error] += 1
+        next
+      end
 
-            $logger.info "Successfully processed Record# #{record['id']}"
-            @process_statuses[:success] += 1
-        }
+      $logger.info("Successfully processed Record# #{record['id']}")
+      @process_statuses[:success] += 1
     end
+    $kinesis_client.push_records
```
This is a nice simplification of this method and reflects how things currently work in the ruby util package.

Unfortunately I think we need to complicate things in order to allow is to retry records.

If a record fails Avro validation then we want to log it and skip it: the record is invalid according to our schema and that won't change so we need to move past it. But if it _passes_ validation and fails when it's written to kinesis then we need to retry it because it _should_ succeed in the future (Paul helped clarify this for me).

This means a few things for how this works I think:
- The `push_records` method in the utils gem needs to be updated to track/store records that errored/failed
- This needs a way to read from that store/array and re-processes these. Either at the end of each batch or at the end of execution.
- It could mean that the structure of this method/class changes to better represent the state as we need it, we don't need to be committed to this way of doing things

mwbenowitz commented on spec/sierra_batch_spec.rb:
```diff
@@ -1,49 +1,41 @@
+require_relative "../lib/sierra_record"
+require_relative "./spec_helper"
 
 describe SierraBatch do
```
I think this may have to change depending on the decisions we make above

mwbenowitz reviewed (commented):
These overall updates look really good, I appreciate the detail in standardizing things and generally cleaning up this repo.

Given some conversations we've had about handling errors, retries and related behavior I think we might need to rework the `SierraBatch` class a bit to provide a more stable processor. With `put_records` being implemented in the utility gem we need a way to detect which records were unable to be put into the target stream and retry them.

Right now we are leaving the specific handling of sending batches to the utility and that might need to shift to this codebase. In addition I think the utility will need to be updated to either store or return errored messages to the invoking application (how those are returned exactly I think is something that should be discussed). Once we have those we can have a method (or a means of re-invoking the original method recursively) that will consume those messages until none error.

None of this is a knock on this PR! This is just something that I've realized that we probably should implement as it would be a major improvement over how I originally conceived of this function (as are all of the improvements being made here). Happy to discuss any of the ideas if there are better ones (or disagreement with this approach)

charmingduchess commented:
@danamansana 
In Sierra Record/Sierra Batch
- changed error and success logging to reflect batch processing
- added retry count to sierra record 
- added calls to retry failed records

mwbenowitz commented on lib/sierra_record.rb:
```diff
@@ -20,12 +21,20 @@ def encode_and_send_to_kinesis
     end
+
     $kinesis_client.push_records
+    @retry_count.times { $kinesis_client.retry_failed_records }
```
could you add some commenting here about this logic? It seems elegant but I'm unfamiliar with the `.times` method. For this method in general to lay out that:

1. it processes everything once, accumulating any errors in the client
2. Retries those failed messages `n` times
3. Logs the ids of any failed messages

mwbenowitz reviewed (approved):
This looks good! If you don't mind adding some comments to the `encode_and_send_to_kinesis` that would be great. It contains the core logic of this class so I think it would be great to make sure that is all explicit (whether for future us or another maintainer)

charmingduchess commented:
Added comments requested [here](https://github.com/NYPL/sierraUpdatePollerV2/pull/19#pullrequestreview-866359774)
